scripts/taxonomy/GetTaxonomyTable.pl

```diff
@@ -0,0 +1,115 @@
+
+use strict;
+use warnings;
+use Getopt::Long;
+use Cwd qw(abs_path);
+use File::Spec::Functions qw(catfile);
+use Bio::EnsEMBL::Compara::DBSQL::DBAdaptor;
+
+$| = 1;
+
+# Parameters:
+my $help = 0;
+# URL to the Compara database
+my $compara_url;
+# Output directory where the taxonomy table will be saved
+my $out_dir = Cwd::abs_path();
+# Handle input arguments
+GetOptions("help"          => \$help,
+           "compara_url=s" => \$compara_url,
+           "outdir=s"      => \$out_dir
+          ) or die("Error in command line arguments\n");
+die "Error in command line arguments -compara_url " .
+    "mysql://user:pass\@host:port/db_name [-outdir /your/directory/]"
+    if (!$compara_url);
+
+# Create new Compara database adaptor and get the GenomeDB adaptor
+my $compara_dba = Bio::EnsEMBL::Compara::DBSQL::DBAdaptor->new(
+    -url => $compara_url, -species => 'Multi');
+my $genome_db_adaptor = $compara_dba->get_GenomeDBAdaptor();
+# Get all the Genome DBs
+my $list_ref_of_gdbs = $genome_db_adaptor->fetch_all();
+# Declare the taxonomic ranks of interest (sorted)
+my @target_ranks = qw(class superorder order suborder family subfamily species);
+# Create the taxonomy template: if a rank is missing, assign "N/A"
+my %rank_template = map {$_ => "N/A"} @target_ranks;
+# Default output file name
+my $out_file = "taxonomy_table.tsv";
+# Recursive function to flatten the taxonomy hash into an array with one string
+# per species' taxonomy (ranks separated by "\t"), and each rank sorted in
+# alphabetical order
+sub flatten {
+    my ($in_hash, $out_array, $already_flatted) = @_;
+    # The first call will have only two arguments
+    $already_flatted = "" unless(defined $already_flatted);
+    for my $key (sort keys %$in_hash) {
+        my $value = $in_hash->{$key};
+        if (ref $value eq 'HASH') {
+            flatten($value, $out_array, $already_flatted . $key . "\t");
+        } else {
+            push(@{$out_array}, $already_flatted . $key);
+        }
+    }
+}
+# Store all the taxonomic information in a mutidimensional hash, following the
+# same order as in @target_ranks
+my $taxonomy_db = {};
+foreach my $ref (@{$list_ref_of_gdbs}) {
+    my $node;
+    # There are some $ref that may not have 'taxon_id', so we need to handle the
+    # exception, report the troubling $ref and move to the next one
+    eval {
+        $node = $ref->taxon();
+    } or do {
+        warn "Undefined taxon_id for '", $ref->name(), "'.\n";
+        next;
+    };
+    # Copy taxonomy template and fill it in
+    my %taxonomy = %rank_template;
+    while ($node->name() ne 'root') {
+        my $rank = $node->rank();
+        if (defined $rank && exists $taxonomy{$rank}) {
+            $taxonomy{$rank} = $node->name();
+        }
+        $node = $node->parent();
+    }
+    # Replace species name by GenomeDB name
+    if (grep(/^species$/, @target_ranks)) {
+        $taxonomy{'species'} = $ref->name();
+    }
+    # Push the template into the taxonomy hash
```

Comment:
Here is a very important part of the script that requires more explanation / better naming.
At this stage, you could just print the line immediately:
```print $file join("\t", map {$taxonomy{$_}} @target_ranks), "\n"; ```
but the output wouldn't be sorted by clades

What you're building in `$taxonomy_db` is essentially the NCBI tree that links all those genomes, but restricted to the allowed ranks. You then use that to sort the genomes according to the taxonomy, and you need to introduce `flatten`. `$taxonomy_db` in fact allows the grouping that Emily was showing in her diagram, e.g. only print "Mammalia" once

1) If we only care about ordering the genomes, you could simplify this by sorting the genome_dbs by the `left_index` of their NCBITaxon object (see https://en.wikipedia.org/wiki/Nested_set_model for an explanation of the "left" and "right" indices. They are used in all our tree tables). Maybe pre-sort the list of genome_dbs and print them on the fly. It's be clearer

2) But since you have the structure that allows the grouping, maybe you could add an option to display it ? e.g. instead of
```
Aves    Neognathae      Anseriformes    N/A     Anatidae        Anatinae        anas_platyrhynchos_platyrhynchos
Aves    Neognathae      Anseriformes    N/A     Anatidae        Anserinae       anser_brachyrhynchus
Aves    Neognathae      Charadriiformes N/A     Scolopacidae    N/A     calidris_pugnax
Aves    Neognathae      Charadriiformes N/A     Scolopacidae    N/A     calidris_pygmaea
```
print
```
Aves    Neognathae      Anseriformes    N/A     Anatidae        Anatinae        anas_platyrhynchos_platyrhynchos
"       "               "               "       "               Anserinae       anser_brachyrhynchus
"       "               Charadriiformes N/A     Scolopacidae    N/A     calidris_pugnax
"       "               "               "       "               "       calidris_pygmaea
```